src/internal/getlogs/matchlines.go

```diff
@@ -0,0 +1,200 @@
+package getlogs
+
+import (
+	"bufio"
+	"encoding/json"
+	"io"
+	"regexp"
+	"time"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errutil"
+)
+
+// MatchLines implements LineMatcher interface
+type MatchLines struct {
+	matchers [](func(line []byte, fields *LogFields) bool)
+}
+
+type LogFields struct {
+	Ts           time.Time
+	Level        string
+	ProjectName  string
+	PipelineName string
+	JobId        string
+	WorkerId     string
+	DatumId      string
+	Master       bool
+	User         bool
+}
+
+func (m *MatchLines) Match(sourceLines io.Reader, returnMatch func([]byte) error) error {
+
+	line := bufio.NewScanner(sourceLines)
+	for line.Scan() {
+		logBytes := line.Bytes()
+
+		logFields := new(LogFields)
+		json.Unmarshal(logBytes, logFields)
+
+		for _, matcher := range m.matchers {
+			if !matcher(logBytes, logFields) {
+				continue
+			}
+			if err := returnMatch(logBytes); err != nil {
+				if errors.Is(err, errutil.ErrBreak) {
+					return nil
+				}
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (m *MatchLines) MatchTime(fromTime time.Time, untilTime time.Time) error {
+	if fromTime.After(untilTime) {
+		return errors.Errorf("invalid parameters, start time %t is after end time %t", fromTime, untilTime)
+	}
+	m.matchers = append(m.matchers, func(line []byte, fields *LogFields) bool {
+		if fields == nil {
+			return false
+		}
+		if fields.Ts.Before(fromTime) || fields.Ts.After(untilTime) {
+			return false
+		}
+		return true
+	})
+	return nil
+}
+
+func (m *MatchLines) MatchRegex(regexStr string, negate bool) error {
+	regex, err := regexp.Compile(regexStr)
+	if err != nil {
+		return errors.Errorf("compile error for regex %q, %q", regexStr, err.Error())
+	}
+	m.matchers = append(m.matchers, func(line []byte, _ *LogFields) bool {
```

Comment:
I think that this should match against each field of the log entry, rather than against the JSON encoding of the log entry. In particular, `"Ts` should probably match against the string `"Ts` within a particular log entry, not every single log entry (assuming that the timestamp field really is named `Ts`, and hence encoded as `{"Ts": …}`

Reply:
Both regex matching on the raw line and on fields are valid use cases - typically the default is to match strings and regexes on raw log lines and provide a way to match fields - I think we can add regexes bound to fields later